substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java

```diff
@@ -0,0 +1,172 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.jfr.traceid;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.oracle.svm.jfr.JfrRuntimeAccess;
+import jdk.jfr.internal.JVM;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+public class JfrTraceId {
+
+    private static final long MaxJfrEventId = 403;
+
+    public static final long BIT = 1;
+    public static final long META_SHIFT = 8;
+    public static final long TRANSIENT_META_BIT = (BIT << 3);
+    public static final long TRANSIENT_BIT = (TRANSIENT_META_BIT << META_SHIFT);
+    public static final long SERIALIZED_META_BIT = (BIT << 4);
+    public static final long SERIALIZED_BIT = (SERIALIZED_META_BIT << META_SHIFT);
+
+    private static final int TRACE_ID_SHIFT = 16;
+
+    // Epoch stuff
+    private static final long USED_BIT = 1;
+    private static final int EPOCH_1_SHIFT = 0;
+    private static final int EPOCH_2_SHIFT = 1;
+    private static final long USED_EPOCH_1_BIT = USED_BIT << EPOCH_1_SHIFT;
+    private static final long USED_EPOCH_2_BIT = USED_BIT << EPOCH_2_SHIFT;
+
+    private static final long JDK_JFR_EVENT_SUBKLASS = 16;
+    private static final long JDK_JFR_EVENT_KLASS = 32;
+    private static final long EVENT_HOST_KLASS = 64;
+
+    private static final AtomicLong threadCounter = new AtomicLong(1);
+
+    private static JfrTraceIdMap getTraceIdMap() {
+        return ImageSingletons.lookup(JfrRuntimeAccess.class).getTraceIdMap();
+    }
+
+    public static void tag(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        map.setId(clazz, (id & ~0xff) | (bits & 0xff));
+    }
+
+    public static boolean predicate(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        return (id & bits) != 0;
+    }
+
+    public static void setUsedThisEpoch(Class<?> clazz) {
+        tag(clazz, JfrTraceIdEpoch.getInstance().thisEpochBit());
+    }
+
+    public static boolean isUsedThisEpoch(Class<?> clazz) {
+        return predicate(clazz, TRANSIENT_BIT | JfrTraceIdEpoch.getInstance().thisEpochBit());
+    }
+
+    public static long getTraceIdRaw(Class<?> clazz) {
+        JfrTraceIdMap map = getTraceIdMap();
+        assert map != null;
+        return getTraceIdMap().getId(clazz);
+    }
+
+    public static long getTraceId(Class<?> clazz) {
+        long traceid = getTraceIdRaw(clazz);
+        return traceid >>> TRACE_ID_SHIFT;
+    }
+
+    public static long load(Class<?> clazz) {
+        return JfrTraceIdLoadBarrier.load(clazz);
+    }
+
+    private static void tagAsJdkJfrEvent(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        assert id != -1;
```

Comment:
Assertions like that one can be removed because you added the assert to `JfrTraceIdMap.getId()`.